compiler/resolution/interfaceResolution.cpp

```diff
@@ -0,0 +1,457 @@
+  FnSymbol* instantiated = required->copy(&fml2actDup);
+  instantiated->addFlag(FLAG_RESOLVED); // this is white lie
+  for_formals(formal, instantiated) {
+    TypeSymbol* fml = formal->type->symbol;
+    if (Symbol* act = fml2act.get(fml))  //wass - taken care by above copy() ?
+      formal->type = toTypeSymbol(act)->type;
+  }
+  return instantiated;
+}
+
+
+static void populateInterfaceFunctionMap(FnSymbol* fn) {
+  INT_ASSERT(fn->interface2actuals.empty()); // first time resolving this fn
+  //wass make this work with gcc 4.8
+  fn->interface2actuals.resize(fn->interfaceConstraints.length);
+  int indx = 0;
+
+  std::vector<FnSymbol*> funs;
+  // For each required function of each interface, make an instantiation of
+  // its signature visible to the function body.
+  for_alist(ieExpr, fn->interfaceConstraints) {
+    ImplementsExpr* ie = toImplementsExpr(ieExpr);
+    InterfaceSymbol* isym = ie->implInterfaceSym();
+    SymbolMap& i2a = fn->interface2actuals[indx++];
+    SymbolMap fml2act;
+    buildAndCheckFormals2Actuals(isym, ie, "interface constraint", fml2act);
+    form_Map(SymbolMapElem, elem, isym->ifcSymbols) {
+      FnSymbol* required = toFnSymbol(elem->key);
+      FnSymbol* instantiated = instantiateRequiredFn(required, fml2act);
+//wass todo:
+// By placing instantiated->defPoint into fn->body, 'fn' will be made visible
+// (by buildVisibleFunctionMap()) throughout fn->body. TODO: also make it
+// visible in 'where' clause. So addVisibleFunctionsToScope(fn->body, funs)
+// will not be needed; it will actually make this function appear twice
+// in the list of resolution candidates, which is bad.
+// Alternatively, use addVisibleFunctionsToScope and do not create
+// instantiated->defPoint. But then overloadSetsOK() will break because
+// overloadSetModule invokes getModule() on 'instantiated'.
+      fn->body->insertAtHead(new DefExpr(instantiated));
+      funs.push_back(instantiated); //wass - need this?
+      i2a.put(required, instantiated);
+    }
+  }
+
+  // CG TODO: we could add the functions in 'funs' to visibleFunctionMap
+  // by calling addVisibleFunctionsToScope(fn->body, funs)
+  // If we end up not calling it, then revert changes in VisibleFunctions.cpp.
+}
+
+
+/*********** resolveConstrainedGenericFun ***********/
+
+// If it a CG function and has not been resolved yet, resolve it.
+void resolveConstrainedGenericFun(FnSymbol* fn) {
+  if (fn->isResolved()) return;
+  if (! fn->isConstrainedGeneric()) return;
+  cgprint("resolving CG function early %s  %s\n", symstring(fn), debugLoc(fn));
+
+  populateInterfaceFunctionMap(fn);
+  // Seems like we do not need fn->setGeneric(false) for resolveFunction()
+  // NB we want 'fn' to be generic later, when checking isApplicable()
+  resolveFunction(fn);
+
+  for_alist(expr, fn->constrainedTypes)
+    toDefExpr(expr)->sym->addFlag(FLAG_GENERIC);
+
+  for (int indx = 0; indx < (int)fn->interface2actuals.size(); indx++) {
```

Comment:
put a comment like "Remove the functions for the interface that we used during resolution"

does this mean that the calls in the constrained generic will be resolved to functions that aren't in tree anymore? Is that a problem?